docs/design/auditing-containment.md

```diff
@@ -0,0 +1,168 @@
+
+The verifier will then attempt to download that erasure share from the contained node.
+From there, the following cases can occur:
+- Passes audit ->  `pendingAudit` removed from the ContainmentDB  -> stats updated in overlaycache/satelliteDB
+- Fails audit -> `pendingAudit` is removed from the ContainmentDB -> stats updated in overlaycache/satelliteDB
+- Refuses to respond again -> the `pendingAudit`'s `reverifyCount` is incremented in the ContainmentDB ->
+    - -> If that count exceeds the reverify limit, then the node’s stats will be updated to reflect an audit failure and the node will be removed from the ContainmentDB
+    - -> If the count does not exceed the reverify limit, the `pendingAudit` will remain in the ContainmentDB
+
+```
+type pendingAudit struct {
+    nodeID       storj.NodeID
+    pieceID      storj.PieceID
+    pieceNum     int
+    stripeIndex  int
+    shareSize    int64
+    expectedShareHash []byte
+    reverifyCount   int // number of times reverify has been attempted
+}
+
+type ContainmentDB interface {
+  Contain(ctx context.Context, pendingAudit pendingAudit) error
+  Get(ctx context.Context, nodeID pb.NodeID) error
+  ReverifyFail(ctx context.Context, nodeID pb.NodeID) error
+  ReverifySuccess(ctx context.Context, nodeID pb.NodeID) error
+}
+```
+
+`pendingAudit` entries are created and added to the ContainmentDB when a node seems to be online and opens a connection with the satellite to be audited (this happens within the existing Verify function in the audit package), but then refuses to send the requested erasure share.
+
+Additionally, the satellite should make sure to empty the ContainmentDB of certain pending audits when segments are deleted.
+
+## Rationale
+
+Originally I had the idea to have a separate service iterate through and check the ContainmentDB.
+However, after talking with JT it seemed unnecessary to have another service for this functionality that could be added to the existing verifier code.
+I think we also don't want to pound the contained nodes with reverification audits.
+They should just be reverified whenever they pass through the audit service normally.
+
+## Implementation
+
+#### 1. Uncooperative nodes are documented in the ContainmentDB:
+
+By "uncooperative," I mean that the auditor will be able to successfully DialNode (within pkg/audit/verifier.go), but errors will occur at
+```
+downloader, err := ps.Download(timedCtx, limit.GetLimit(), offset, int64(shareSize))
+```
+or
+```
+_, err = io.ReadFull(downloader, buf)
+```
+Once these errors have occurred and we know that a `pendingAudit` entry should be made in the ContainmentDB, we should use infectious’s `f.Decode` to generate the original stripe.
+Then we should use infectious’s `f.EncodeSingle` where we input the stripe and output the missing share. We then make a hash of that missing share to be saved to the ContainmentDB as `ExpectedShareHash`.
```

Comment:
This is cool.  What happens to containment if we don't have enough nodes to Decode()?

Reply:
Hmm that's a good question. Currently in the existing code, an error will be returned from `verifier.Verify` if less than the required number of shares are available (this happens when `f.Correct` is called).
It seems not good to just return an error without "containing" this node. But it also seems bad to "contain" the node without saving its `ExpectedShareHash`.

Maybe we could make a case where we save the segment's path to the `pendingAudit` entry and then when reverify is attempted, it could get the latest pointer and try to download again (with hopefully more nodes available)? But maybe that is too convoluted. I would be interested in hearing other thoughts.